### PR TITLE
Add EIP: A Protocol for ERC20 Tokens Which Released Based On Predefined Periods Controlled by Smart Contracts

### PR DESCRIPTION
**Pull request by erc4377**
A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts, enhancing the ERC repository. 
The smart contract allows token released based on predefined periods, enabling lock token balances and gradually release tokens.
**ATTENTION: ERC-RELATED PULL REQUESTS NOW OCCUR IN [ETHEREUM/ERCS](https://github.com/ethereum/ercs)**

--

When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

File: EIPS/eip-4377.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
+---
+eip: 4377
```
Comment (abcoathup) [inline EIPS/eip-4377.md:2]:
```suggestion
eip: 
```
ERC numbers are issued by editors & associates.  They are now sequential.

Once you have created a PR in the ERC repo a number will be manually issued.

**Comment by abcoathup** [comment]
To create an ERC (application layer) please close this PR and create a new PR in the ERC repo: https://github.com/ethereum/ERCs

**Comment by abcoathup** [comment]
@erc4377 please close this PR now that you have created a PR in the ERC repo: https://github.com/ethereum/ERCs/pull/341

File: EIPS/eip-4377.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+title: A Protocol for ERC20 Tokens Which Released Based On Predefined Periods Controlled by Smart Contracts
```
Comment (gaudren) [inline EIPS/eip-4377.md:3]: Titles must be 45 characters or less. Please rework it to be more succinct.

File: EIPS/eip-4377.md [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-03-08
+---
+
+## Abstract
+
+A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts, enhancing the ERC repository. 
+The smart contract allows token released based on predefined periods, enabling lock token balances and gradually release tokens.
```
Comment (gaudren) [inline EIPS/eip-4377.md:16]: Re-states same information as the line above:

```suggestion
```

File: EIPS/eip-4377.md [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+## Specification
+
+This protocol introduces improvements in the following areas:
+
+1. ERC20 Tokens and Linear Release Functionality.
+2. Online Representation of ERC20 Balances.
+3. Lock-Up Functionality for ERC20 Tokens.
+4. "Sorter" Algorithm.
+5. Low-Cost Swap Algorithm.
```
Comment (gaudren) [inline EIPS/eip-4377.md:34]: This should be in the Motivation section.

File: EIPS/eip-4377.md [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+## Rationale
+
+There are four major improvements in this protocol:
+
+* ERC20 Tokens and Linear Release Functionality.
+ canReleaseAmount = total * (currentTimestamp - startTime) / duration;
```
Comment (gaudren) [inline EIPS/eip-4377.md:41]:
```suggestion
 `canReleaseAmount = total * (currentTimestamp - startTime) / duration;`
```

File: EIPS/eip-4377.md [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+category: Core
+created: 2024-03-08
+---
+
+## Abstract
+
+A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts, enhancing the ERC repository. 
+The smart contract allows token released based on predefined periods, enabling lock token balances and gradually release tokens.
+
+Our team has introduced a novel “Sorter” mechanism in this protocol, which automatically consolidates multiple transfers in smart contracts. This feature ensures token release based on different transfer times while minimizing gas fees significantly.
```
Comment (gaudren) [inline EIPS/eip-4377.md:18]:
```suggestion
This proposal introduces a novel “Sorter” mechanism in this protocol, which automatically consolidates multiple transfers in smart contracts. This feature ensures token release based on different transfer times while minimizing gas fees significantly. 
```

File: EIPS/eip-4377.md [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+discussions-to: https://ethereum-magicians.org/t/erc4377-a-protocol-for-erc20-tokens-which-released-based-on-predefined-periods-controlled/19305
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-03-08
+---
+
+## Abstract
+
+A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts, enhancing the ERC repository.
```
Comment (gaudren) [inline EIPS/eip-4377.md:15]:
```suggestion
A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts. 
```

File: EIPS/eip-4377.md [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+## Abstract
+
+A protocol for ERC20 tokens which released based on predefined periods controlled by smart contracts, enhancing the ERC repository. 
+The smart contract allows token released based on predefined periods, enabling lock token balances and gradually release tokens.
+
+Our team has introduced a novel “Sorter” mechanism in this protocol, which automatically consolidates multiple transfers in smart contracts. This feature ensures token release based on different transfer times while minimizing gas fees significantly. 
+Furthermore, the protocol supports customizable features such as:
+
+Customized White List: Specified addresses or contracts are exempted from the above features and locking restrictions, allowing this protocol to be widely and properly applied. For example, Uniswap contracts, treasury addresses.... etc.
```
Comment (gaudren) [inline EIPS/eip-4377.md:21]:
```suggestion
Customized Whitelist: Specified addresses or contracts are exempted from the above features and locking restrictions, allowing this protocol to be widely and properly applied. Examples include Uniswap contracts, and treasury addresses.
```

You might want to explain how these examples are relevant.

File: EIPS/eip-4377.md [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+
+## Motivation
+The motivation behind this protocol is to enrich the ERC repository. By allowing ERC20 tokens to have custom linear release schedules, lock-up mechanisms, and whitelisting, this protocol can find applications in scenarios such as MEME tokens, games, DAO organizations, and options trading.
```
Comment (gaudren) [inline EIPS/eip-4377.md:24]:
```suggestion
By allowing ERC-20 tokens to have custom linear release schedules, lock-up mechanisms, and whitelisting, this protocol can find applications in scenarios such as MEME tokens, games, DAO organizations, and options trading.
```

File: EIPS/eip-4377.md [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+
+This protocol introduces improvements in the following areas:
+
+1. ERC20 Tokens and Linear Release Functionality.
+2. Online Representation of ERC20 Balances.
+3. Lock-Up Functionality for ERC20 Tokens.
+4. "Sorter" Algorithm.
+5. Low-Cost Swap Algorithm.
+
+## Rationale
```
Comment (gaudren) [inline EIPS/eip-4377.md:36]: This doesn't explain the choices you have made for this EIP, and code block should generally be included in the Reference Implementation. Use this section to help the reader understand why you have done something in plain English.

File: EIPS/eip-4377.md [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+
+Our team has introduced a novel “Sorter” mechanism in this protocol, which automatically consolidates multiple transfers in smart contracts. This feature ensures token release based on different transfer times while minimizing gas fees significantly. 
+Furthermore, the protocol supports customizable features such as:
+
+Customized White List: Specified addresses or contracts are exempted from the above features and locking restrictions, allowing this protocol to be widely and properly applied. For example, Uniswap contracts, treasury addresses.... etc.
+
+## Motivation
+The motivation behind this protocol is to enrich the ERC repository. By allowing ERC20 tokens to have custom linear release schedules, lock-up mechanisms, and whitelisting, this protocol can find applications in scenarios such as MEME tokens, games, DAO organizations, and options trading.
+
+## Specification
```
Comment (gaudren) [inline EIPS/eip-4377.md:26]: You don't have any specifications. What is an implementer supposed to do?

File: EIPS/eip-4377.md [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+
+## Copyright
+
+Copyright and related rights waived via [CC0](../LICENSE.md).
+
+## Citation
+
+Please cite this document as:
+
+Roy Davies(roydj2021@gmail.com), Ali Kaya (ali.kayatur747@gmail.com), "ERC-4377: A Protocol for ERC20 Tokens Which Released Based On Predefined Periods Controlled by Smart Contracts [DRAFT]," Ethereum Improvement Proposals, no. 4377, March 2024. [Online serial]. Available: https://eips.ethereum.org/EIPS/eip-4377.
```
Comment (gaudren) [inline EIPS/eip-4377.md:360]:
```suggestion
```

This section will be generated automatically when rendered.

**Review by gaudren — CHANGES_REQUESTED** [review]
